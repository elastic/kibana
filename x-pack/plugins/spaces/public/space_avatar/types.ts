/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import type { Space } from '../../common';

/**
 * Properties for the SpaceAvatar component.
 */
export interface SpaceAvatarProps {
  /** The space to represent with an avatar. */
  space: Partial<Space>;

  /** The size of the avatar. */
  size?: 's' | 'm' | 'l' | 'xl';

  /** Optional CSS class(es) to apply. */
  className?: string;

  /**
   * When enabled, allows EUI to provide an aria-label for this component, which is announced on screen readers.
   *
   * Default value is true.
   */
  announceSpaceName?: boolean;

  /**
   * Whether or not to render the avatar in a disabled state.
   *
   * Default value is false.
   */
  isDisabled?: boolean;
}
