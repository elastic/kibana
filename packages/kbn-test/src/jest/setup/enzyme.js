/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });
