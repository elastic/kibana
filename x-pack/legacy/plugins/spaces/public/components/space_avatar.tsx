/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { EuiAvatar, isValidHex } from '@elastic/eui';
import React, { SFC } from 'react';
import { getSpaceColor, getSpaceInitials, MAX_SPACE_INITIALS } from '../../common';
import { Space } from '../../common/model/space';

interface Props {
  space: Partial<Space>;
  size?: 's' | 'm' | 'l' | 'xl';
  className?: string;
  announceSpaceName?: boolean;
}

export const SpaceAvatar: SFC<Props> = (props: Props) => {
  const { space, size, announceSpaceName, ...rest } = props;

  const spaceName = space.name ? space.name.trim() : '';

  const spaceColor = getSpaceColor(space);

  return (
    <EuiAvatar
      type="space"
      data-test-subj={`space-avatar-${space.id}`}
      name={spaceName}
      {...(!announceSpaceName && {
        // provide empty aria-label so EUI doesn't try to provide its own
        'aria-label': '',
        'aria-hidden': true,
      })}
      size={size || 'm'}
      initialsLength={MAX_SPACE_INITIALS}
      initials={getSpaceInitials(space)}
      color={isValidHex(spaceColor) ? spaceColor : ''}
      {...rest}
    />
  );
};

SpaceAvatar.defaultProps = {
  announceSpaceName: true,
};
