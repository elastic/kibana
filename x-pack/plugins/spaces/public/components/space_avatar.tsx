/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { EuiAvatar } from '@elastic/eui';
import React from 'react';
import { getSpaceColor, getSpaceInitials, MAX_SPACE_INITIALS } from '../../common';
import { Space } from '../../common/model/space';

interface Props {
  space: Partial<Space>;
  size?: 's' | 'm' | 'l' | 'xl';
  className?: string;
}

export const SpaceAvatar = (props: Props) => {
  const { space, size, ...rest } = props;

  const spaceName = space.name ? space.name.trim() : '';

  return (
    <EuiAvatar
      type="space"
      data-test-subj={`space-avatar-${space.id}`}
      name={spaceName}
      size={size || 'm'}
      initialsLength={MAX_SPACE_INITIALS}
      initials={getSpaceInitials(space)}
      color={getSpaceColor(space)}
      {...rest}
    />
  );
};
