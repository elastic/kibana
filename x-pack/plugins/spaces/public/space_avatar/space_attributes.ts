/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { VISUALIZATION_COLORS } from '@elastic/eui';

import type { Space } from 'src/plugins/spaces_oss/common';

import { MAX_SPACE_INITIALS } from '../../common';

// code point for lowercase "a"
const FALLBACK_CODE_POINT = 97;

/**
 * Determines the color for the provided space.
 * If a color is present on the Space itself, then that is used.
 * Otherwise, a color is provided from EUI's Visualization Colors based on the space name.
 *
 * @param {Space} space the space.
 */
export function getSpaceColor(space: Partial<Space> = {}) {
  const { color, name = '' } = space;

  if (color) {
    return color;
  }

  const firstCodePoint = name.codePointAt(0) || FALLBACK_CODE_POINT;

  return VISUALIZATION_COLORS[firstCodePoint % VISUALIZATION_COLORS.length];
}

/**
 * Determines the initials for the provided space.
 * If initials are present on the Space itself, then that is used.
 * Otherwise, the initials are calculated based off the words in the space name, with a max length of 2 characters.
 *
 * @param {Space} space the space.
 */
export function getSpaceInitials(space: Partial<Space> = {}) {
  const { initials, name = '' } = space;

  if (initials) {
    return initials;
  }

  const words = name.split(' ');

  const numInitials = Math.min(MAX_SPACE_INITIALS, words.length);

  words.splice(numInitials, words.length);

  return words.map((word) => word.substring(0, 1)).join('');
}

/**
 * Determines the avatar image for the provided space.
 *
 * @param {Space} space the space.
 */
export function getSpaceImageUrl(space: Partial<Space> = {}) {
  const { imageUrl } = space;

  if (imageUrl) {
    return imageUrl;
  }

  return '';
}
