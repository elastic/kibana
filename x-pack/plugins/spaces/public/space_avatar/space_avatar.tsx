/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';

import type { SpaceAvatarProps } from 'src/plugins/spaces_oss/public';

export const getSpaceAvatarComponent = async (): Promise<React.FC<SpaceAvatarProps>> => {
  const { SpaceAvatarInternal } = await import('./space_avatar_internal');
  return (props: SpaceAvatarProps) => {
    return <SpaceAvatarInternal {...props} />;
  };
};
