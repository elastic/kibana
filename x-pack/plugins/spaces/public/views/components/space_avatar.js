/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import {
  EuiAvatar
} from '@elastic/eui';
import { MAX_SPACE_INITIALS, getSpaceInitials, getSpaceColor } from '../../../common';

export const SpaceAvatar = ({ space, size, ...rest }) => {
  return (
    <EuiAvatar
      type="space"
      name={space.name || ''}
      size={size || "m"}
      initialsLength={MAX_SPACE_INITIALS}
      initials={getSpaceInitials(space)}
      color={getSpaceColor(space)}
      {...rest}
    />
  );
};

SpaceAvatar.propTypes = {
  space: PropTypes.object.isRequired,
  size: PropTypes.string,
};
