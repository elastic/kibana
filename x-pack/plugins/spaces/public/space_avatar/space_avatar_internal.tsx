/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import type { EuiAvatarProps } from '@elastic/eui';
import { EuiAvatar, isValidHex } from '@elastic/eui';
import type { FC } from 'react';
import React from 'react';

import type { Space } from 'src/plugins/spaces_oss/common';

import { MAX_SPACE_INITIALS } from '../../common';
import { getSpaceColor, getSpaceImageUrl, getSpaceInitials } from './space_attributes';

interface Props {
  space: Partial<Space>;
  size?: 's' | 'm' | 'l' | 'xl';
  className?: string;
  announceSpaceName?: boolean;
  /**
   * This property is passed to the underlying `EuiAvatar` component. If enabled, the SpaceAvatar will have a grayed out appearance. For
   * example, this can be useful when rendering a list of spaces for a specific feature, if the feature is disabled in one of those spaces.
   * Default: false.
   */
  isDisabled?: boolean;
}

export const SpaceAvatarInternal: FC<Props> = (props: Props) => {
  const { space, size, announceSpaceName, ...rest } = props;

  const spaceName = space.name ? space.name.trim() : '';

  const spaceColor = getSpaceColor(space);

  const spaceInitials = getSpaceInitials(space);

  const spaceImageUrl = getSpaceImageUrl(space);

  const avatarConfig: Partial<EuiAvatarProps> = spaceImageUrl
    ? { imageUrl: spaceImageUrl }
    : { initials: spaceInitials, initialsLength: MAX_SPACE_INITIALS };

  return (
    <EuiAvatar
      type="space"
      data-test-subj={`space-avatar-${space.id}`}
      name={spaceName}
      {...(!announceSpaceName && {
        // provide empty aria-label so EUI doesn't try to provide its own
        'aria-label': '',
        'aria-hidden': true,
      })}
      size={size || 'm'}
      color={isValidHex(spaceColor) ? spaceColor : ''}
      {...avatarConfig}
      {...rest}
    />
  );
};

SpaceAvatarInternal.defaultProps = {
  announceSpaceName: true,
};
