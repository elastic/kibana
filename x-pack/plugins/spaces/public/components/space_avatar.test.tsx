/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { SpaceAvatar } from './space_avatar';

test('renders without crashing', () => {
  const wrapper = shallow(<SpaceAvatar space={{ name: '', id: '' }} />);
  expect(wrapper).toMatchSnapshot();
});

test('renders with a space name entirely made of whitespace', () => {
  const wrapper = shallow(<SpaceAvatar space={{ name: '      ', id: '' }} />);
  expect(wrapper).toMatchSnapshot();
});
