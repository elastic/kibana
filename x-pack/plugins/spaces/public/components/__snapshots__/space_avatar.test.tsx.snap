// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with a space name entirely made of whitespace 1`] = `
<EuiAvatar
  color="#DB1374"
  data-test-subj="space-avatar-"
  initials=""
  initialsLength={2}
  name=""
  size="m"
  type="space"
/>
`;

exports[`renders without crashing 1`] = `
<EuiAvatar
  color="#BFA180"
  data-test-subj="space-avatar-"
  initials=""
  initialsLength={2}
  name=""
  size="m"
  type="space"
/>
`;
